_posts/2017-11-21-Installing-Jenkins-with-PowerShell-DSC.md

```diff
@@ -0,0 +1,745 @@
+
+1. We need to define the AllNodes.NodeName that we referred in the script. 
+2. We also need to call give the parameters we want to the configuration. 
+3. Finally we start the configuration with Start-DscConfiguraiton 
+
+```powershell
+$ConfigData = @{
+    AllNodes = 
+    @(
+        @{
+            NodeName = "LocalHost"
+        }
+    )
+}
+
+JENKINS_CI -JenkinsPort 8080 -ConfigurationData $ConfigData
+Start-DscConfiguration -Path .\JENKINS_CI -Wait -Verbose -Force
+```
+
+## All the needed dependencies
+
+We will need a bunch of stuff to be able to build a .NET application. 
+
+* Java for the Jenkins
+* Visual Studio and MSBuild for all the build targets and compilation. 
+* NodeJs for the front end compilation.
+* Git for grabbing few more resources
+* Firefox for OWASP Zap 
+* Notepad++ for comfort
+* Nuget.exe (I like to have a fixed version of this)
+* Python for ... things 
+* Jenkins for the actual goal 
+* Zap for web application security testing automation
+
+Here is the installation of the stuff above. So we insert more stuff under our $AllNodes.NodeName. 
+
+```powershell
+# Install Visual Studio, todo: optional features (F#) with param --includeOptional
+cChocoPackageInstaller installVisualStudio
+{
+	Name = "visualstudio2017professional"
+	DependsOn = "[cChocoInstaller]installChoco"
+}
+
+# Install Visual Studio Web tools 
+cChocoPackageInstaller installVisualStudioWebWorkload
+{
+	Name = "visualstudio2017-workload-netweb"
+	Params = "--includeOptional"
+	DependsOn = "[cChocoInstaller]installChoco","[cChocoPackageInstaller]installVisualStudio"
+}
+
+# Install NuGet
+File installNuget 
+{
+	DestinationPath = "C:\tools\nuget\nuget.exe"
+	SourcePath = (Join-Path $InstallConfDirectory "nuget.exe")
+	Ensure = "Present"
+	Type = "File"
+	Checksum = "modifiedDate"
+	Force = $true
+	MatchSource = $true
+}
+```
+
+I think that the above script is quite self-explanatory although I made it shorter for easier reading. The full script can be found [here](https://github.com/solita/powershell-dsc-jenkins/blob/master/jenkins_dsc.ps1) We have bunch of resources, most of them depends on something earlier. Afterwards we want to have all those resources in place. Only odd thing is in the installVisualStudioWebWorkload where I add includeOptional parameter for the choco. This is to get F# installed on the target machine as well. It does not come by default. If you need to dig on to those additional options in nuget packages you should head to chocolateys web page and investigate the packages you need to figure out if there is any customization options.
+
+## Installing custom made stuff
+
+We have something in our GitHub repositories that we would like to use with Jenkins. Those are marvelous libraries jmeter-perfotrator and powershell-zap. We can fetch the git repository directly into Jenkins server and whatever what is there with the help of just installed git client. Here is an example resource for that.
```

Comment:
"We can fetch the git repository directly into Jenkins server and whatever what is there with the help of just installed git client" -> "We can fetch files from a git repository directly into Jenkins server with the help of the git client we just installed."